### PR TITLE
Added field support to JSON serializer

### LINKED ISSUE #876: JsonSerializer should support field as well as properties
_Updated by @layomia:  I'm pasting the API approved in https://github.com/dotnet/runtime/issues/34558. Please see that issue for review notes and further discussion._

```C#
namespace System.Text.Json
{
    public partial class JsonSerializerOptions
    {
        public bool IgnoreReadOnlyFields { get; set; }
        public bool IncludeFields { get; set; }
    }
}
namespace System.Text.Json.Serialization
{
    [AttributeUsage(AttributeTargets.Property |
                    Attributes.Field, AllowMultiple = false)]
    public sealed class JsonIncludeAttribute : JsonAttribute
    {
        public JsonIncludeAttribute();
    }
}
```

<details>
<summary>
Original proposal (click to view)
</summary>

There is no way to serialize and deserialize fields using `JsonSerializer`. 

While public fields are [generally not recommended](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/fields), they are used in .NET itself (see value tuples) and by users.

This feature was scoped out of v1 (3.x) due to lack of time, as we prioritized supporting simple POCOs with public properties.

We elected to have an opt-in model for field support because it would be a breaking change to support them by default, and also because public fields are generally not recommended. Other serializers, including `Newtonsoft.Json`, `Utf8Json`, and `Jil`, support this feature by default.
</details>

### PR DESCRIPTION
**Pull request by YohDeadfall**
Fixes #876, continues #2192.

@layomia This is reworked version of what I did before. It has no tests (in progress), but I would like to get your comments as soon as possible. And yep, there were unanswered questions in the previous PR. Hope you will find some time to address them (:

### CODE REVIEW

**Comment by layomia** [comment]
Addressing questions from https://github.com/dotnet/runtime/pull/2192:

> Should we rename all parameters, fields and properties of type PropertyInfo to something having member as a part of the name?

I think the renaming done in the current PR is fine i.e. using `MemberInfo` rather than`PropertyInfo`. To keep the size of the PR smaller, `JsonPropertyInfo` and co should not be renamed.

> Having a text template to generate class/struct for visibility and fields would improve code coverage and remove duplication.

I'd be supportive of having a template to do this if you want to take this on in this PR, depending on the implementation. However, for the purpose of driving this feature in, I'd be satisfied with interspersing fields in the existing test objects.

Wrt. to coverage for the new code, we should have tests for all serializer features that apply to properties e.g. `JsonExtensionData`, `JsonIgnore` (including ignoring per-filed), `JsonInclude`, `JsonConverterAttribute`, `JsonPropertyName` etc. This is where the duplication may be more pronounced.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (layomia) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:525]: `JsonIncludeAttribute` should also have fields as a target. This is helpful when `JsonSerializerOptions.IncludeFields`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs:8]:
```suggestion
    /// When placed on a property, field, or type, specifies the converter type to use.
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs:12]:
```suggestion
    /// When placed on a property or field, the specified converter will always be used.
```

Let's use "property or field" rather than "member" here and for similar occurrences, since this is not applicable to other members like methods and indexers.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:158]: We need to extend the property visibility tests added in https://github.com/dotnet/runtime/pull/32107 and https://github.com/dotnet/runtime/pull/36936 to include collisions between field names, and collisions between property and field names.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:106]:
```suggestion
                            const BindingFlags bindingFlags =
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:137]: FWI you'll likely have to react to https://github.com/dotnet/runtime/pull/36936 here.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:158]: Any way to extract the following logic into a method and share between fields and properties above?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:451]: We need to test fields in the `InvalidTests` test class.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs:130]: We need to add the `IgnoreReadonlyFields` property specified in https://github.com/dotnet/runtime/issues/876#issue-537934155 and update `DetermineSerializationCapabilities()` above accordingly.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:71]: Why extra braces here?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs:224]: Looks like there are some CI failures relating to the delegate creation. I'll let you address those, then I'll circle back to the member accessor changes.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs:48]: No need to change this for now.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs:318]: No need to change this for now. Let's move this up, where "RuntimePropertyType" is.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:158]: Adding to the property/field cache needs to be dependent on an if-check for `options.IncludeFields`, or the presence of `JsonInclude` on the field.

**Review by layomia — COMMENTED** [review]
Thanks for the PR. There are a couple new `JsonSerializerOptions` properties from https://github.com/dotnet/runtime/issues/876#issue-537934155 yet to be added. I brought up a few test cases to add. `MemberAccessor` review is pending on a fix for the CI failures.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:525]: Missed this, thanks!

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:158]: It's totally possible to have only loops invoking a generic method accepting two non-allocating delegates or usual delegates to check visibility and to create a property.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:71]: To align it with getters/setters. Remove it?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: I was thinking about something along these lines to avoid some of the code duplication. Note, I included the checks for whether we should actually include the field:

```cs
for (Type? currentType = type; currentType != null; currentType = currentType.BaseType)
{
    const BindingFlags bindingFlags =
        BindingFlags.Instance |
        BindingFlags.Public |
        BindingFlags.NonPublic |
        BindingFlags.DeclaredOnly;

    foreach (PropertyInfo propertyInfo in currentType.GetProperties(bindingFlags))
    {
        // Ignore indexers
        if (propertyInfo.GetIndexParameters().Length > 0)
        {
            continue;
        }

        bool isPublic = propertyInfo.GetMethod?.IsPublic == true || propertyInfo.SetMethod?.IsPublic == true;
        if (isPublic)
        {
            CacheMember(propertyInfo, propertyInfo.PropertyType, currentType, cache, options);
        }
        else
        {
            if (JsonPropertyInfo.GetAttribute<JsonIncludeAttribute>(propertyInfo) != null)
            {
                ThrowHelper.ThrowInvalidOperationException_JsonIncludeOnNonPublicInvalid(propertyInfo, currentType);
            }

            // Non-public properties should not be included for (de)serialization.
        }
    }

    foreach (FieldInfo fieldInfo in currentType.GetFields(bindingFlags))
    {
        bool hasJsonInclude = JsonPropertyInfo.GetAttribute<JsonIncludeAttribute>(fieldInfo) != null;
        bool shouldInclude = options.IncludeFields || hasJsonInclude;

        if (!fieldInfo.IsPublic)
        {
            if (hasJsonInclude)
            {
                ThrowHelper.ThrowInvalidOperationException_JsonIncludeOnNonPublicInvalid(fieldInfo, currentType);
            }

            // Non-public fields should not be included for (de)serialization.
        }
        else if (shouldInclude)
        {
            CacheMember(fieldInfo, fieldInfo.FieldType, currentType, cache, options);
        }
    }
}

private void CacheMember(
    MemberInfo memberInfo,
    Type memberType,
    Type parentClassType,
    Dictionary<string, JsonPropertyInfo> cache,
    JsonSerializerOptions options)
{
    JsonPropertyInfo jsonPropertyInfo = AddProperty(memberInfo, memberType, parentClassType, options);
    Debug.Assert(jsonPropertyInfo.NameAsString != null);

    // If the JsonPropertyNameAttribute or naming policy results in collisions, throw an exception.
    if (!JsonHelpers.TryAdd(cache, jsonPropertyInfo.NameAsString, jsonPropertyInfo))
    {
        JsonPropertyInfo other = cache[jsonPropertyInfo.NameAsString];

        if (other.ShouldDeserialize == false && other.ShouldSerialize == false)
        {
            // Overwrite the one just added since it has [JsonIgnore].
            cache[jsonPropertyInfo.NameAsString] = jsonPropertyInfo;
        }
        else if (other.MemberInfo?.Name != jsonPropertyInfo.MemberInfo?.Name &&
            (jsonPropertyInfo.ShouldDeserialize == true || jsonPropertyInfo.ShouldSerialize == true))
        {
            // Check for name equality is required to determine when a new slot is used for the member.
            // Therefore, if names are not the same, there is conflict due to the name policy or attributes.
            ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(Type, jsonPropertyInfo);
        }
        // else ignore jsonPropertyInfo since it has [JsonIgnore] or it's hidden by a new slot.
    }
}
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:71]: Yes.

**Comment by layomia** [comment]
Ping @YohDeadfall. Marking as ready-for review since we are close to the implementation we want. There's still some feedback above to be addressed.

**Comment by YohDeadfall** [comment]
Already addressed some of them, but not pushed (running tests). Am trying to narrow how the changes cause CI failures.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs:224]: Not sure that the failures are caused by this changes, because the exceptions says about the parameter mismatch. The problem comes from broken JSON property info creation I assume.

**Comment by YohDeadfall** [comment]
@layomia could you help me? After rebasing the PR and installing the latest .NET, am unable to build .NET anymore because of:

```
.nuget\packages\microsoft.dotnet.build.tasks.sharedframework.sdk\5.0.0-beta.20278.4\targets\framework.dependency.targets(445,5): error MSB3073: The command "\dotnet\runtime\artifacts\bin\coreclr\Windows_NT.x64.Debug\sharedFramework\crossgen.exe @\dotnet\runtime\artifacts\obj\netcoreapp\Debug\net5.0\win-x64\crossgen\System.Runtime.WindowsRuntime.rsp" exited with code -2146233031.
```

This blocks me from fixing the issue with constructors.

**Comment by YohDeadfall** [comment]
Cleaning up `artifacts` helped.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (YohDeadfall) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:285]: @layomia this line caused failures because `CreateIgnoredPropertyPlaceholder` uses `byte` for placeholder properties. The easiest way to fix it is to do `switch` on `memberInfo` and get its type, but I don't like it.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:285]: Why don't you like it? I think switching on `memberInfo` is fine, since we previously got it from `propertyInfo`, not `jsonPropertyInfo` (perhaps for the same reason).

Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:285]: Not obvious that an ignored JSON property doesn't represent the corresponding CLR property, plus code bloating because we should repeat type determination when there should be one entry point.

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: We can revisit this.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs:224]: Yep, this causes problems too. For example, `TestClassWithFieldsHavingCustomConverter` has property `DerivedCustomer` of type `DerivedCustomer`. The member accessor correctly generates methods using the `DerivedCustomer`, but a callee requested method for type `Customer`.

How this happens? `JsonConverter<T>` is responsible for a property creation for now, but its `T` can be less derived than the actual property type.

**Comment by layomia** [comment]
@YohDeadfall anything blocking progress on this? I can look into the test failures if needed. Preview 7 snaps on Monday, and we are trying to get this in soon.

**Comment by YohDeadfall** [comment]
Just time blocks it accidentally, nothing more. I aware of the failures and know how to fix them.

**Comment by layomia** [comment]
Looks like CI failures are due to the merge conflicts.

**Comment by layomia** [comment]
Are you holding off for green CI before adding tests?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs]: Are there still issues with this?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:71]: Since we don't check for `JsonIncludeAttribute` like we do for properties, I assume based on a prior review that we don't support non-public fields. cc @layomia

**Comment by steveharter** [comment]
Please add the two test classes from the prior PR @ https://github.com/dotnet/runtime/pull/2192 and be sure to hook them up to `TestData.ReadSuccessCases` and `TestData.WriteSuccessCases`. Thanks

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:69]: We do check for `JsonIncludeAttribute` to include public fields "per-field". This is useful when the global option, `options.IncludeFields`, is `false`. Non-public fields are not supported; `InvalidOperationException` is thrown if the attribute is placed on a non-public field (to prevent a breaking change if we have non-public member support in the future).

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs:69]: We should `Debug.Assert(fieldInfo.IsPublic)`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:233]: Why is this passed by ref?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:240]: We've made a few changes recently to make sure we have the correct semantics for property name collisions. We should be sure to add tests to cover collisions between field names, and also between property and field names. Basically, we should extend this test suite: https://github.com/dotnet/runtime/blob/8fd53646fc1a099db797d0177b7d9d1eac581939/src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs#L14-L431

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:292]: Is there any measurable perf benefit to using `Unsafe.As` here over a regular cast?

**Review by layomia — COMMENTED** [review]
CI is green 👍  We just need the tests now :)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:233]: Because it's `null` by default and assigned only in this method on line 268.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:292]: One less cast. This trick is used by `ValueTask` for example, spans have it too.

Personally I would like to avoid these repeatable lines of code, but there is no other way to get the true property type because placeholders use `byte`. I told you about my concerns on that before.

A good way to avoid this is to have `JsonPlaceholderPropertyInfo` which won't be generic, will implement all required stuff to represent a property as it is, but which will throw on any attempt to read or write a value. In that case it will be fine to use `JsonPropertyInfo.DeclaredPropertyType` or `RuntimePropertyType`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs]: Won't work on .NET Framework, will add an extension method to fix the problem for old TFMs, For modern platforms, it won't exist because of conditional compilation.

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs
Comment (YohDeadfall) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs:140]: @layomia What should be here to fix the Windows build?

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs:140]: I think we should just revert to just the helper you had earlier. I don't think using this simple built-in public helper is worth the multi-targeting complexity.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:358]: Copy/paste error?

Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:358]: What do you mean?

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:358]: It appeared that this chunk of code was repeated. Either my eyes tricked me, or it was resolved.

Comment (YohDeadfall) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:358]: No, that code hasn't changed, but it really looks like the previous lines (:

Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs:358]: AH, got it.

**Comment by layomia** [comment]
@YohDeadfall, will you have some time soon to check on the failures here?

**Comment by YohDeadfall** [comment]
Am on vacation till Saturday, so on Sunday will return back and finish the feature.

**Review by layomia — COMMENTED** [review]
LGTM when these comments are addressed.

File: src/libraries/System.Text.Json/tests/Serialization/SpanTests.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/tests/Serialization/SpanTests.cs:34]: Yeah, I pushed tests as is before going to vacation. Thought that would have some time, but had other interests there.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs
Comment (YohDeadfall) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:484]: Added some new tests, but not finished yet. For naming they already exist.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs:14]: nit: property or field

File: src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs:59]:
```suggestion
        public static void ExtensionFieldNotUsed()
```

**Review by layomia — APPROVED** [review]
LGTM - thanks for your contribution, @YohDeadfall.

**Comment by layomia** [comment]
Will address minor clean up.

**Pull request merged**
